r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs

```diff
@@ -179,211 +569,445 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
+        // ``
+        //      0 <= L[i]*R[i]  < (num_add(L) + 1) * (num_add(R) + 1) * 2^{2 * bits_per_limb[i]}, 
+        // ``
+        // and hence 
+        // ``
+        //     0 <= prod_limb[j] < 
+        //          num_limbs * (num_add(L) + 1) * (num_add(R) + 1) * 2^bits_per_prod_limb[j],
+        // ``
+        // where `bits_per_prod_limb[j] = 2 * bits_per_limb` for all except the most significant
+        // limb, and `bits_per_prod_limb[0] = 2 * bits_per_limb[0]`.
+        // Hence we may set 
+        // ``
+        //      num_add(product) =  num_limbs * (num_add(L) + 1) * (num_add(R) + 1) - 1. 
+        // ``
+        let result = NonNativeFieldMulResultGadget {
+            limbs: prod_limbs,
+            num_add_over_normal_form: &num_add_bound - &BigUint::one(),
+            simulation_phantom: PhantomData,
+        };
 
-                let mut x = FpGadget::<ConstraintF>::zero(cs.ns(|| format!("alloc x {}", c)))?;
-                for (i, (var, c_pow)) in self_reduced.limbs.iter().zip(c_pows.iter()).enumerate() {
-                    let mul_result = var.mul_by_constant(
-                        cs.ns(|| format!("self var * c_pow[{}]{}", i, c)),
-                        &c_pow,
-                    )?;
-                    x.add_in_place(cs.ns(|| format!("x + mul result {},{}", c, i)), &mul_result)?;
-                }
+        debug_assert!(
+            result.check(),
+            "mul_without_prereduce(): result failed on check()"
+        );
 
-                let mut y = FpGadget::<ConstraintF>::zero(cs.ns(|| format!("alloc y {}", c)))?;
-                for (i, (var, c_pow)) in other_reduced.limbs.iter().zip(c_pows.iter()).enumerate() {
-                    let mul_result = var.mul_by_constant(
-                        cs.ns(|| format!("other var * c_pow[{}]{}", i, c)),
-                        &c_pow,
-                    )?;
-                    y.add_in_place(cs.ns(|| format!("y + mul result {},{}", c, i)), &mul_result)?;
-                }
+        Ok(result)
+    }
 
-                let mut z = FpGadget::<ConstraintF>::zero(cs.ns(|| format!("alloc z {}", c)))?;
-                for (i, (var, c_pow)) in prod_limbs.iter().zip(c_pows.iter()).enumerate() {
-                    let mul_result = var.mul_by_constant(
-                        cs.ns(|| format!("prod var * c_pow[{}]{}", i, c)),
-                        &c_pow,
-                    )?;
-                    z.add_in_place(cs.ns(|| format!("z + mul result {},{}", c, i)), &mul_result)?;
-                }
+    /// For advanced use, multiply and output the intermediate representations (without reduction)
+    /// This intermediate representations can be added with each other, and they can later be
+    /// reduced back to the `NonNativeFieldGadget`.
+    /// Assumes that
+    /// ``
+    ///     2 * bits_per_limb + surfeit' <= CAPACITY - 2,
+    /// ``
+    /// where
+    /// ``
+    ///      surfeit' = log(num_limbs^2 * (num_add(L)+1) + 1).
+    /// ``
```

Comment:
Specify in the inline comment: The constant is reduced in normal form: this is why it does not appear num_add(R)..